### DISCUSSION TITLE
OASDIFF Circular Dependency problem.

### DISCUSSION THREAD

**Discussion by darshil594** [Tufin/oasdiff#387] [Q&A]
I'm currently using OASDiff to track changes in a Swagger REST API, and I've encountered an issue with circular dependencies in my JSON schema. Unfortunately, I can't modify the JSON schema to resolve these circular dependencies.

`Error: failed to load base spec from "FILENAME.json" with kin-openapi bug found: circular schema reference not handled - #/A/B -> #/A/C -> #/A/C -> #/A/C -> #/A/C -> #/A/C -> #/A/C. Where A , B and C are classname.
`
s there any way to configure OASDiff to ignore this circular schema reference error and generate the output report anyway? I need to track API changes, and resolving the circular dependencies in the JSON schema is not currently an option.

**Comment by reuvenharrison** [comment]
Hi @darshil594 
Can you share the specs and cmd-line that replicate this problem?

**Reply by tejaswinni51** [reply]
Hi @reuvenharrison, I'm also facing a similar issue with our swagger.json file, This is the command we used 

**oasdiff diff tejaswini767.json tejaswini7113.json -f HTML** 

**Output :**
Error: failed to load base spec from "tejaswini767.json" with kin-openapi bug found: circular schema reference not handled - #/definitions/ApiReplicationScheduleList -> #/definitions/ApiReplicationSchedule -> #/definitions/ApiReplicationCommand -> #/definitions/ApiCommand -> #/definitions/ApiCommand -> #/definitions/ApiCommand -> #/definitions/ApiCommand -> #/definitions/ApiCommand -> #/definitions/ApiCommandList -> #/definitions/ApiCommand

The JSON file contains references of ApiCommandList in ApiCommand and ApiCommand is inherited in ApiCommandList class.

 ```
 "ApiCommandList" : {
      "type" : "object",
      "title" : "ApiCommandList",
      "allOf" : [
        {
          "$ref" : "#/definitions/ApiListBase"
        },
        {
          "properties" : {
            "items" : {
              "description" : "",
              "type" : "array",
              "items" : {
                "$ref" : "#/definitions/ApiCommand"
              }
            }
          }
        }
      ],


```

 ```
"ApiCommand" : {
      "type" : "object",
      "title" : "ApiCommand",
      "properties" : {
        "id" : {
          "description" : "The command ID.",
          "type" : "integer"
        },
        "name" : {
          "description" : "The command name.",
          "type" : "string"
        },
        .
        . 
        .
        .
        .,
        "children" : {
          "description" : "List of child commands. Only available in the full view.\n<p>\nThe list contains only the summary view of the children.",
          "$ref" : "#/definitions/ApiCommandList"
        },
        .
        .
        .
        .
      },
```

can you suggest/help us?

**Comment by darshil594** [comment]
> Hi @darshil594 Can you share the specs and cmd-line that replicate this problem?

Hi @reuvenharrison , I have the similar schema. And using the same command line oasdiff diff file1.json file2.json --format html > op.html.

There are many **#ref** used within the JSON, and the JSON file itself is quite large. Is there a possibility of encountering a heap memory issue due to this? If so, can we modify the heap memory allocation to address it?

**Reply by reuvenharrison** [reply]
Hi @darshil594 and @tejaswini22199 
I would love to assist but to do so I need your help to replicate the error on my dev machine.
Could you please share the full specs and cmd-line that I can run to get the same error?
If you like, you can send them to me privately on Slack:

https://join.slack.com/t/oasdiff/shared_invite/zt-1wvo7wois-ttncNBmyjyRXqBzyg~P6oA

**Comment by reuvenharrison** [comment]
This is just a guess, but try setting `--max-circular-dep` to a higher value, for example 15 and see if the problem dissappeers.

**Reply by darshil594** [reply]
Hi @reuvenharrison ,

Thank you for your response. I have attempted to set the circular depth to values such as 15, 150, and even 1500. However, the problem persists, and there are still numerous circular dependencies. Even when I set the depth to 15000, the error persists. Seems like there are many circular dependency. 

but Converting the Swagger from 2.0 to 3.0 using the Swagger Convert API resolves our problem.

**Reply by reuvenharrison** [reply]
oasdiff only supports OpenAPI v3 so I don't really understand how this process worked, but if your happy with the result that's great.
